libraries/chain/include/eosio/chain/hotstuff.hpp

```diff
@@ -12,10 +12,46 @@ namespace eosio::chain {
 
    using hs_bitset = boost::dynamic_bitset<uint32_t>;
 
+   static digest_type get_digest_to_sign(const block_id_type& block_id, uint8_t phase_counter, const fc::sha256& final_on_qc){
+      digest_type h1 = digest_type::hash( std::make_pair( block_id, phase_counter ) );
+      digest_type h2 = digest_type::hash( std::make_pair( h1, final_on_qc ) );
+      return h2;
+   }
+
    inline uint64_t compute_height(uint32_t block_height, uint32_t phase_counter) {
       return (uint64_t{block_height} << 32) | phase_counter;
    }
 
+   struct view_number {
+      view_number() : _block_height(0), _phase_counter(0) {}
+      explicit view_number(std::pair<uint32_t, uint8_t> data) :
+         _block_height(data.first),
+         _phase_counter(data.second)
+      {
+      }
+      
+      explicit view_number(uint32_t block_height, uint8_t phase_counter) :
+         _block_height(block_height),
+         _phase_counter(phase_counter)
+      {
+      }
+
+      auto operator<=>(const view_number&) const = default;
+
+      uint32_t block_height() const  { return _block_height; }
+      uint8_t  phase_counter() const { return _phase_counter; }
+
+      uint64_t    to_uint64_t() const { return compute_height(_block_height, _phase_counter);   }
+      std::string to_string()   const { return std::to_string(_block_height) + "::" + std::to_string(_phase_counter); }
+
+      uint64_t get_key() const { return get_view_number().to_uint64_t(); };
+
+      view_number get_view_number() const { return view_number{ block_height(), phase_counter() }; };
```

Comment:
This is confusing to me. Why does this exist? How is it just not `view_number a = ...; view_number b = a;` instead of `view_number b = a.get_view_number();` ?